Extension/src/Utility/Process/streams.ts

```diff
@@ -0,0 +1,566 @@
+
+        // carry over any partial line from before
+        if (this.#partial) {
+            incoming[0] = this.#partial + incoming[0];
+            this.#partial = undefined;
+        }
+
+        for (let line of incoming) {
+            line = this.trimTrailingWhitespace ? line.trimEnd() : line;
+
+            if (this.readStream) {
+                // call notifyReading quickly, so that we don't block the stream
+                this.readStream(line);
+            }
+
+            if (this.streamRead) {
+                // if we have a reading event, then we queue it up
+                done.push(this.streamRead(line).then((text) => !is.cancelled(text) ? this.push(text !== undefined ? text : line) : undefined));
+            } else {
+                // otherwise, we just push the line into the buffer.
+                this.push(line);
+            }
+        }
+
+        if (done.length) {
+            // if we have any pending emitted events, wait for them to finish before signaling that we have new lines
+            void Promise.all(done).then(() => this.changed.resolve());
+        } else {
+            // signal that we have new lines
+            this.changed.resolve();
+        }
+    }
+
+    /** clears the buffer */
+    clear() {
+        this.trim();
+    }
+
+    /**
+     * Trim elements from the front of the buffer.
+     * @param count the number of elements to trim (defaults to trimming the whole buffer.)
+     */
+    trim(keepMaxLines = 0) {
+        // figure out where the new head should be
+        const newHead = Math.max(this.head, this.last - keepMaxLines);
+
+        // if we are actually trimming (and the head should move forward), then fill the elements with undefined
+        if (newHead > 0) {
+            this.#buffer.fill(undefined as unknown as string, this.head, newHead);
+
+            // set the new head position
+            this.#head = newHead;
+        }
+    }
+
+    /** returns a copy of the entire line buffer */
+    all() {
+        const result = this.#buffer.slice(this.head);
+        if (this.#partial) {
+            result.push(this.#partial);
+        }
+        return result;
+    }
+}
+
+export class ReadWriteLineStream extends ReadableLineStream {
+    #encoder = new TextEncoder();
+    protected writeable: Writable;
+
+    setWriteEvent(event: (text: string) => Promise<EventStatus | string>) {
+        this.streamWrite = event;
+    }
+    setWriteNotifier(notifier: (text: string) => void) {
+        this.wroteStream = notifier;
+    }
+    protected wroteStream?: (text: string) => void;
+    protected streamWrite?: (text: string) => Promise<EventStatus | string>;
+
+    constructor(stream: Duplex);
+    constructor(readable: Readable, writeable: Writable);
+    constructor(readable: Readable | Duplex, writeable?: Writable) {
+        super(readable);
+        this.writeable = writeable || readable as Writable;
+
+        this.writeable.on('error', (_error) => {
+            /*

```

Comment:
Is this supposed to commented out or removed? If it should be kept, can you add a comment explaining when to uncommen it?

Reply:
added comment